### PR TITLE
pow accuracy improvement and port new algorithm to Blackhole

### LINKED ISSUE #19821: TTNN pow op calculates incorrect results for input > 1
**Describe the bug**
ttnn.pow(9,2) != 81

**To Reproduce**
import torch
import ttnn
import numpy as np

torch.manual_seed(0)

device_id = 0
device = ttnn.open_device(device_id=device_id)
ttnn.enable_program_cache(device)

torch_a = torch.randn((1), dtype=torch.float32) #bfloat16)
torch_b = torch.randn((1), dtype=torch.float32) #bfloat16)
torch_a[0] = (-9)
torch_b[0] = (2)

a = ttnn.from_torch(torch_a, layout=ttnn.TILE_LAYOUT, device=device)
b = ttnn.from_torch(torch_b, layout=ttnn.TILE_LAYOUT, device=device)
a_torch = a.cpu().to_torch()
b_torch = b.cpu().to_torch()
#np_a  = a_torch.cpu().numpy()
#np_b  = b_torch.cpu().numpy()

output = ttnn.pow(a, b);
torch_output = torch.pow(torch_a,torch_b)

output = ttnn.to_layout(output, ttnn.ROW_MAJOR_LAYOUT)

print("Printing ttnn tensor")
print(f"shape: {output.shape}")

output_torch = output.cpu().to_torch()
#out_np = np.power(np_a, np_b)

#print(a_torch[0], "^", b_torch[0], " torch: ", torch_output[0], " ttnn:  ", output_torch[0], " numpy: ", out_np[0])
print(a_torch[0], "^", b_torch[0], " torch: ", torch_output[0], " ttnn:  ", output_torch[0])

ttnn.close_device(device)

**Expected behavior**
9^2 == 81

**Screenshots**

**Please complete the following environment information:**
- OS: [e.g. Ubuntu 20.04]
- Version of software (eg. commit)

**Additional context**

### LINKED ISSUE #23529: POW output mismatch against CPU reference implementation
### Describe the bug

When computing tensor power operations (e.g., tensor ** 2), the TT backend produces results that differ from the reference CPU computation. 

### Steps to reproduce the issue

```
test_vals = torch.tensor([9.0], dtype=torch.bfloat16)
expected = torch.tensor([81.0], dtype=torch.bfloat16)  # 9^2
actual = torch.ops.aten.pow.Tensor_Scalar(test_vals.to(device), 2)
```

### Expected behavior
Expected result: 81.0 (bfloat16)
Actual result: 78.0 (bfloat16)

### LINKED ISSUE #25815: Make pow function round outputs to-nearest bfloat16 instead of truncating it.
PR #25025 significantly improved the accuracy of `pow`.

Right now, some values are within 1 ULP of error from golden outputs. This 1ULP error sometimes happen because storing float32 data from LRegs to bfloat16 DST registers truncates values. https://github.com/tenstorrent/tt-llk/issues/494

To fix this, we want to round output to bfloat16 before writing it to DST register. This means adding a `is_fp32_dest_acc_en ` template parameter to pow-related LLK functions and performing this round-to-nearest when false (similar to sqrt-related functions).

### LINKED ISSUE #25816: Port new pow implementation to Blackhole
Due to SFPI-related bug/behavior that leads to incorrect results (#25647 ), the new power function (#25025 ) has not been ported to blackhole. 

Once the bug is fixed (or a workaround is found), the new power function should be enabled for blackhole.

### PR DESCRIPTION
**Pull request by nmauriceTT**
### Ticket
#25815

### Problem description
`LRegs` in the SFPU use 32-bits data (e.g. float32). When the DST register data type is set to bfloat16, LRegs values are truncated upon storage. 
This leads to small 1ULP errors such as `9**2 = 80.5` rather than 81 (float32 output would have been 80.819 and should have been rounded to 81 if using bfloat16). 

A proposed solution to this issue was suggested here. It is similar to what's used with `recip_tile` .

Moreover, a typo has been found in the magical numbers of `pow`: one constant should be `0x560e` rather than `0x560`. This is a relatively small floating point number, hence why it was not caught before. 

Beyond these two issues, there was a bug on Blackhole that prevented the implementation of the new pow algorithm ( #25647 ) but it has been fixed recently ( #26467 ) . 

### What's changed
- Add a `is_fp32_dest_acc_en` template parameter to binary pow LLK. The pow approximation uses this parameter to properly round float32 values in `LRegs` to the nearest float16 rather than truncating it upon storing to DST registers. 
- Fix typo in magic number in `pow` function (improve accuracy by ~20% on average)
- Port pow algorithm to Blackhole (SFPI bug has been fixed by #26467 )
- Remove blackhole skips in `test_pow.py`

### Accuracy improvement

The improvement in accuracy seems to be by a factor of around 2. 

As with PR https://github.com/tenstorrent/tt-metal/pull/25025, I've measured and plotted the accuracy of `pow` exhaustively on every bfloat16 combination. 

I've plotted the accuracy results for current `pow` (left) and `pow` with proper rounding. 
In this case:  darker = better (yellow means it's either infinite or NaN). 

As the heatmap show: rounding-to-nearest improves accuracy overall, and let us reach a perfect ULP for some small values. 

466109367-45d0cc8d-3ec4-42a7-8afd-deb9c66153a1
pow21f_round_nearest

Sampling the functions on a few values gives us the following table:

#### bfloat16

| x | y | torch | ttnn.pow (current) | tttn.pow (proposed) |
| - | - | - | - | - |
| 9 | 2 | 81 | 80.5 | 81 |
| 100000 | 1.7984 | 9.6050e+08 | 9.6050e+08 | 9.6050e+08 |
| 5 | 3 | 125 | 124.5 | 125 |

The new function provides exacts results for these 3 inputs, which was not case before. 

#### float32 (not affected by rounding)

| x | y | torch | ttnn.pow (current) | tttn.pow (proposed) |
| - | - | - | - | - |
| 9 | 2 | 81 | 80.819 | 80.940 |
| 100000 | 1.7984 | 9.6050e+08 | 9.80734592+08 | 9.81990528e+08 |
| 5 | 3 | 125 | 124.961 | 125.116 |

#### Execution time overhead

This correct rounding mode adds one instruction `float_to_fp16b` / [`SFPSTOCHRND`](https://github.com/tenstorrent/tt-isa-documentation/blob/main/WormholeB0/TensixTile/TensixCoprocessor/SFPSTOCHRND_FloatFloat.md) for kernels that use it. 

From my testing (profiling while looping 100000 times on the LLK ), on Wormhole, the execution time is the same as current version.  
On Blackhole, the algorithm is ~5% faster than 'legacy' implementation, while being significantly more accurate (similar to what was seen on Wormhole in previous PR #25025 )

Note: I've also been working on a PR to add the `is_fp32_dest_acc_en` to most SFPU numerical LLKs (at least for later): #26133. 

### Checklist
- [X] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes : https://github.com/tenstorrent/tt-metal/actions/runs/16805139790 (same failures as `main` & unrelated fail on gather)
- [X] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) : https://github.com/tenstorrent/tt-metal/actions/runs/16830785713
- [X] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable) : https://github.com/tenstorrent/tt-metal/actions/runs/16779407293
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-quick-trigger.yaml) CI passes (if applicable)
- [ ] [TG demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/tg-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [x] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes : https://github.com/tenstorrent/tt-metal/actions/runs/16779382454 (same failures as [main](https://github.com/tenstorrent/tt-metal/actions/runs/16780917269) )
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [X] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h [hunk 1]
```diff
@@ -13,14 +13,193 @@ using namespace sfpi;
...
+
+template <bool APPROXIMATION_MODE, BinaryOp BINOP, int ITERATIONS = 8, bool is_fp32_dest_acc_en = false>
 inline void calculate_sfpu_binary(const uint dst_offset) {
-    _calculate_sfpu_binary_<APPROXIMATION_MODE, BINOP, ITERATIONS>(dst_offset);
+    if constexpr (BINOP == BinaryOp::POW) {
+        for (int d = 0; d < ITERATIONS; d++) {
+            constexpr uint dst_tile_size = 32;
+            sfpi::vFloat in0 = sfpi::dst_reg[0];
+            sfpi::vFloat in1 = sfpi::dst_reg[dst_offset * dst_tile_size];
+            sfpi::vFloat result = 0.f;
```
Comment (amahmudTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h]: Is there any reason to assign result 0.f explicitly since it is going to be overwritten anyways.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h [hunk 2]
```diff
@@ -13,14 +13,193 @@ using namespace sfpi;
...
+template <bool is_fp32_dest_acc_en = false>
+sfpi_inline sfpi::vFloat _sfpu_binary_power_(sfpi::vFloat base, sfpi::vFloat pow) {
+    // The algorithm works in two steps:
+    // 1) Compute log2(base)
+    // 2) Compute base**pow = 2**(pow * log2(base))
+
+    // Step 1: Compute log2(base)
+    // Normalize base to calculation range
+    sfpi::vFloat absbase = setsgn(base, 0);       // set base as positive
+    sfpi::vFloat x = sfpi::setexp(absbase, 127);  // set exp to exp bias (put base in range of 1-2)
```
Comment (amahmudTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h:53]: 127 ? not 23 or something like that ?

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h (hunk 1 above)
Comment (nmauriceTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h]: It was meant to be a sanity check during testing. I've removed it.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h (hunk 2 above)
Comment (nmauriceTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h:53]: (Note: The log computation in this specific section is from the legacy approximation)

In this case, a range reduction to [1; 2] is performed to compute the logarithm. That means that the exponent must be set to 0. Since `setexp` uses bias of -127, we are using 127 to set the exponent.

The '23' constant is used for something else. We are using this number because there are 23 mantissa bits. 
For instance, if the exponent is 23 (i.e. `setexp(x, 127 + 23)`), then the floating point number is necessarily an integer. And in that situation, changing the last mantissa bit means adding 1 or removing 1. 
That's likely it's used in relation to float -> int steps (and why both the `pow` algorithm and the `_float_to_int32_positive_` function are using this 23 value).

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h [hunk 3]
```diff
@@ -13,14 +13,192 @@ using namespace sfpi;
...
+}
+
+template <bool APPROXIMATION_MODE, BinaryOp BINOP, int ITERATIONS = 8, bool is_fp32_dest_acc_en = false>
 inline void calculate_sfpu_binary(const uint dst_offset) {
-    _calculate_sfpu_binary_<APPROXIMATION_MODE, BINOP, ITERATIONS>(dst_offset);
+    if constexpr (BINOP == BinaryOp::POW) {
+        for (int d = 0; d < ITERATIONS; d++) {
+            constexpr uint dst_tile_size = 32;
+            sfpi::vFloat in0 = sfpi::dst_reg[0];
+            sfpi::vFloat in1 = sfpi::dst_reg[dst_offset * dst_tile_size];
```
Comment (dchenTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h:157]: both dst_offset and dst_tile_size is constant, so the multiply needs to be done once. Compiler should optimize it to once. Just make sure, otherwise explictly move to outside of loop.

**Review by dchenTT — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h (hunk 3 above)
Comment (nmauriceTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h:157]: I don't see any `mul` instruction in the output assembly, so it should be good. 

I haven't gone through all the assembly, but I believe `dst_offset * dst_tile_size` is either reduced to a constant (both should be known at compile time), or computed with a bitwise shift (`x * 32 == x << 6`).

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h [hunk 4]
```diff
@@ -13,14 +13,192 @@ using namespace sfpi;
 namespace ckernel {
 namespace sfpu {
 
-template <bool APPROXIMATION_MODE, BinaryOp BINOP, int ITERATIONS = 8>
+// Helper function for _sfpu_binary_power_
+// This function is based on _float32_to_int32_, but expects a positive input, which allows us to optimize
+// away several lines (and make it faster)
+sfpi_inline sfpi::vInt _float_to_int32_positive_(sfpi::vFloat in) {
```
Comment (nvelickovicTT) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h]: We don't have a neat place for these helper functions, but I think it is time we create one.
Otherwise, to reuse these helper functions in other files, requires including this one, which creates bloat. And we already had such an instance with `sfpu_sign_mag_to_twos_comp`.

I would advocate that ckernel_defs.h in TT-LLK is the best place for that, but as a first iteration, creating something like ckernel_helpers.h in tt-metal would be fine as well.

**Review by nvelickovicTT — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h (hunk 4 above)
Comment (nmauriceTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h]: I agree with the bloat problem. In this case, we'll probably use this helper function in both `exp` and `exp2` too. 

That being said, this function is derived off `_float_to_int32_` in which I removed code that handles negative numbers. 
For this specific instance, wouldn't it also make sense to add `_float_to_int32_positive_` next to `_float_to_int32_`, or in a new `ckernel_sfpu_rounding_ops.h` file in tt-metal for now ?

**Review by cmaryanTT — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h (hunk 4 above)
Comment (nmauriceTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h]: I've moved the functions to a new `ckernel_sfpu_rounding_ops.h` in `tt_metal`. 
Let me know if you prefer having it in something helper-specific ( even if rounding-related).

Comment (nvelickovicTT) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_binary.h]: Thanks. Definitely like the separate file approach better. I would just maybe suggest to call it `ckernel_sfpu_conversions.h` and put all these helper functions that do conversions there.
That seems cleanest to me.
